pkg/kubelet/cm/helpers.go

```diff
@@ -0,0 +1,97 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package cm
+
+import (
+	"fmt"
+	"path"
+
+	libcontainercgroups "github.com/opencontainers/runc/libcontainer/cgroups"
+	cgroupfs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
+	libcontainerconfigs "github.com/opencontainers/runc/libcontainer/configs"
+)
+
+// cgroupSubsystems holds information about the mounted cgroup subsytems
+type cgroupSubsystems struct {
+	// Cgroup subsystem mounts.
+	// e.g.: "/sys/fs/cgroup/cpu" -> ["cpu", "cpuacct"]
+	mounts []libcontainercgroups.Mount
+
+	// Cgroup subsystem to their mount location.
+	// e.g.: "cpu" -> "/sys/fs/cgroup/cpu"
+	mountPoints map[string]string
+}
+
+// GetCgroupSubsystems returns information about the mounted cgroup subsystems
+func getCgroupSubsystems() (cgroupSubsystems, error) {
+	// Get all cgroup mounts.
+	allCgroups, err := libcontainercgroups.GetCgroupMounts()
+	if err != nil {
+		return cgroupSubsystems{}, err
+	}
+	if len(allCgroups) == 0 {
+		return cgroupSubsystems{}, fmt.Errorf("failed to find cgroup mounts")
+	}
+
+	//TODO(@dubstack) should we trim to only the supported ones
+	mountPoints := make(map[string]string, len(allCgroups))
+	for _, mount := range allCgroups {
+		for _, subsystem := range mount.Subsystems {
+			mountPoints[subsystem] = mount.Mountpoint
+		}
+	}
+	return cgroupSubsystems{
+		mounts:      allCgroups,
+		mountPoints: mountPoints,
+	}, nil
+}
+
+// getLibcontainerCgroupManager returns libcontainer's cgroups manager
+// object with the specified cgroup configuration
+func getLibcontainerCgroupManager(cgroupConfig *CgroupConfig) (*cgroupfs.Manager, error) {
+	// get cgroup name
+	name := cgroupConfig.Name
+
+	// Extract the cgroup resource parameters
+	resourceConfig := cgroupConfig.ResourceParameters
+	resources := &libcontainerconfigs.Resources{}
+	if resourceConfig.Memory != nil {
+		resources.Memory = *resourceConfig.Memory
+	}
+	if resourceConfig.CpuShares != nil {
+		resources.CpuShares = *resourceConfig.CpuShares
+	}
+	if resourceConfig.CpuQuota != nil {
+		resources.CpuQuota = *resourceConfig.CpuQuota
+	}
+	// Initialize libcontainer's cgroup config
+	libcontainerCgroupConfig := &libcontainerconfigs.Cgroup{
+		Name:      path.Base(name),
+		Parent:    path.Dir(name),
+		Resources: resources,
+	}
+	// get all the mounted subsystems
+	cgroupSubsystems, err := getCgroupSubsystems()
+	if err != nil {
+		return &cgroupfs.Manager{}, fmt.Errorf("Unable to destroy cgroup paths for cgroup %v: %v", name, err)
+	}
+	// Get map of all cgroup paths on the system for the particular cgroup
+	cgroupPaths := make(map[string]string, len(cgroupSubsystems.mountPoints))
+	for key, val := range cgroupSubsystems.mountPoints {
+		cgroupPaths[key] = path.Join(val, name)
+	}
+	return &cgroupfs.Manager{
+		Cgroups: libcontainerCgroupConfig,
+		Paths:   cgroupPaths,
```

Comment:
We need it here only for the destroy() method. We create a new cgroupfs manager object for each method. Because of this when destory() method is called Paths need to be initialized.

I believe in other places the same object  which created the cgroup must be used to destroy the cgroups. In that case it paths is stored when the apply function is called.